hrms/api/roster.py

```diff
@@ -0,0 +1,258 @@
+import frappe
+from frappe import _
+from frappe.utils import add_days, date_diff, random_string
+
+from hrms.hr.doctype.shift_assignment.shift_assignment import ShiftAssignment
+from hrms.hr.doctype.shift_assignment_tool.shift_assignment_tool import create_shift_assignment
+
+
+@frappe.whitelist()
+def get_values(doctype: str, name: str, fields: list) -> dict[str, str]:
+	return frappe.db.get_value(doctype, name, fields, as_dict=True)
+
+
+@frappe.whitelist()
+def get_events(
+	month_start: str, month_end: str, employee_filters: dict[str, str], shift_filters: dict[str, str]
+) -> dict[str, list[dict]]:
+	holidays = get_holidays(month_start, month_end, employee_filters)
+	leaves = get_leaves(month_start, month_end, employee_filters)
+	shifts = get_shifts(month_start, month_end, employee_filters, shift_filters)
+
+	events = {}
+	for event in [holidays, leaves, shifts]:
+		event = group_by_employee(event)
+		for key, value in event.items():
+			if key in events:
+				events[key].extend(value)
+			else:
+				events[key] = value
+	return events
+
+
+@frappe.whitelist()
+def create_shift_assignment_schedule(
+	employee: str,
+	company: str,
+	shift_type: str,
+	status: str,
+	start_date: str,
+	end_date: str | None,
+	days: list[str],
+	frequency: str,
+) -> None:
+	schedule = frappe.get_doc(
+		{
+			"doctype": "Shift Assignment Schedule",
+			"schedule": random_string(10),
+			"frequency": frequency,
+			"days": [{"day": day} for day in days],
+			"status": "Inactive" if end_date else "Active",
```

Comment:
**Status** 'Active' Schedules are ongoing Schedules for which the scheduler will continue to make shift assignments. If an end date has been provided for a Schedule, it will be 'Inactive' because assignments will not need to be created after that date.

**End Date of Last Shift Created** stores the End Date of the latest Shift Assignment of that Schedule as a reference date for the scheduler for creating future shifts. Kind of like 'Process Attendance After' from Auto Attendance.

**Shift Status** is, well, for storing the Status of the Shift Assignment.

Will add necessary descriptions for these.